src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -137,94 +144,177 @@ boolean removeUnreachableBlocks ()
+	 * {@link L2ReadOperand}s may now be referring to {@link L2SemanticValue}s
+	 * that are inaccessible at the point of read, via an {@link L2Register}
+	 * that is still accessible.  Replace the {@link L2SemanticValue} of that
+	 * read to one that's present in every {@link L2WriteOperand} that writes to
+	 * the register.
 	 *
-	 * @return Whether any dead instructions were removed or changed.
+	 * <p>It should be the case that there will always be at least one common
+	 * semantic value among all writes to a register.</p>
 	 */
-	private boolean removeDeadInstructions ()
+	private void updateAllSemanticValuesOfReads ()
 	{
-		boolean anyRemoved = false;
-		final Set<L2Instruction> neededInstructions = findNeededInstructions();
+		final Map<L2Register, L2SemanticValue> favoredSemanticValuesMap =
+			new HashMap<>();
 		for (final L2BasicBlock block : blocks)
 		{
-			final Iterator<L2Instruction> iterator =
-				block.instructions().iterator();
-			while (iterator.hasNext())
+			for (final L2Instruction instruction : block.instructions())
 			{
-				final L2Instruction instruction = iterator.next();
-				if (!neededInstructions.contains(instruction))
+				if (instruction.operation().isPhi())
 				{
-					anyRemoved = true;
-					iterator.remove();
-					instruction.justRemoved();
+					// Don't mess with the semantic values coming into a phi, as
+					// they're always expected to mutually agree.
+					continue;
+				}
+				for (final L2ReadOperand<?> read : instruction.readOperands())
+				{
+					@Nullable L2SemanticValue favoredSemanticValue =
+						favoredSemanticValuesMap.get(read.register());
+					if (favoredSemanticValue == null)
+					{
+						@Nullable Set<L2SemanticValue> intersection =
+							null;
+						for (final L2WriteOperand<?> write :
+							read.register().definitions())
+						{
+							if (intersection == null)
+							{
+								intersection =
+									new HashSet<>(write.semanticValues());
+							}
+							else
+							{
+								intersection.retainAll(write.semanticValues());
+							}
+						}
+						assert intersection != null;
+						assert !intersection.isEmpty();
+						favoredSemanticValue = intersection.iterator().next();
+						favoredSemanticValuesMap.put(
+							read.register(), favoredSemanticValue);
+					}
+					read.updateSemanticValue(favoredSemanticValue);
 				}
 			}
 		}
-		return anyRemoved;
 	}
 
 	/**
 	 * Remove all unreachable blocks and all instructions that don't either have
 	 * a side-effect or produce a value ultimately used by an instruction that
 	 * has a side-effect.
+	 *
+	 * @param dataCouplingMode
+	 *        How to trace data dependencies.
 	 */
-	void removeDeadCode ()
+	void removeDeadCode (
+		final DataCouplingMode dataCouplingMode)
 	{
-		//noinspection StatementWithEmptyBody
-		while (removeUnreachableBlocks() || removeDeadInstructions()) { }
+		while (removeUnreachableBlocks()
+			|| removeDeadInstructions(dataCouplingMode))
```

Comment:
Ah, some instructions can be dead inside live blocks because their writes are not read subsequently. I forgot about this possibility. That answers one of my questions for earlier.

Reply:
Correct.